msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1263,6 +1267,130 @@ public boolean isSharedDevice() {
+            } else {
+                throw asyncResult.getException();
+            }
+        } catch (ExecutionException | InterruptedException e) {
+            throw new MsalClientException(
+                    UNKNOWN_ERROR,
+                    "Unexpected error while generating SHR.",
+                    e
+            );
+        }
+    }
+
+    @Override
+    public void generateSignedHttpRequest(@NonNull final IAccount account,
+                                          @NonNull final PoPAuthenticationScheme popParameters,
+                                          @NonNull final SignedHttpRequestRequestCallback callback) {
+        final GenerateShrCommandParameters cmdParams =
+                CommandParametersAdapter.createGenerateShrCommandParameters(
+                        mPublicClientConfiguration,
+                        mPublicClientConfiguration.getOAuth2TokenCache(),
+                        ((Account) account).getHomeAccountId(),
+                        popParameters
+                );
+        try {
+            final GenerateShrCommand generateShrCommand = new GenerateShrCommand(
+                    cmdParams,
+                    MSALControllerFactory.getAllControllers(
+                            mPublicClientConfiguration.getAppContext(),
+                            mPublicClientConfiguration.getDefaultAuthority(),
+                            mPublicClientConfiguration
+                    ),
+                    new CommandCallback<GenerateShrResult, BaseException>() {
+                        @Override
+                        public void onCancel() {
+                            // Not cancellable
+                        }
+
+                        @Override
+                        public void onError(@NonNull final BaseException error) {
+                            callback.onError(baseExceptionToMsalException(error));
+                        }
+
+                        @Override
+                        public void onTaskCompleted(@NonNull final GenerateShrResult generateShrResult) {
+                            callback.onTaskCompleted(generateShrResult.getShr());
+                        }
+                    },
+                    "1101" // Public API ID
+            );
+
+            // Execute this command silently...
+            CommandDispatcher.submitSilent(generateShrCommand);
+        } catch (final MsalClientException e) {
+            final MsalClientException clientException = new MsalClientException(
+                    UNKNOWN_ERROR,
+                    "Unexpected error while generating SHR.",
+                    e
+            );
+            callback.onError(clientException);
+        }
+    }
+
+    private MsalException baseExceptionToMsalException(@NonNull final BaseException exception) {
+        if (GenerateShrResult.Errors.NO_ACCOUNT_FOUND.equalsIgnoreCase(exception.getErrorCode())) {
+            return new MsalUiRequiredException(
```

Comment:
so basically the user needs to get a token? even though they don't want it

Reply:
Yep. This is by design. The only way to get an SPO cookie to sign is by redeeming a PoP-AT for it -- the `generateShr` proposal is predicated on using the same signing keys for a user as _would have_ been used to sign an AT, even though the AT is not included in the resulting payload.